File: golang-problems/combinations.go
package golang_problems

import "slices"

func combinations(nums []int, k int) [][]int {
	var loop func(int)

	var results [][]int
	var stack = make([]int, 0, k)

	loop = func(start int) {
		if len(stack) == k {
			results = append(results, slices.Clone(stack))
			return
		}

		requested := k - len(stack)
		available := func(i int) int {
			return len(nums) - i
		}
		for i := start; i < len(nums) && requested <= available(i); i++ {
			stack = append(stack, nums[i])
			loop(i + 1)
			stack = stack[:len(stack)-1]
		}
	}

	loop(0)
	return results
}

func combine(n int, k int) [][]int {
	var loop func(int)

	var results [][]int
	var stack []int

	loop = func(start int) {
		if len(stack) == k {
			results = append(results, slices.Clone(stack))
			return
		}

		requested := k - len(stack)
		available := func(i int) int {
			return n - i
		}

		for i := start; i < n && requested <= available(i); i++ {
			stack = append(stack, i+1)
			loop(i + 1)
			stack = stack[:len(stack)-1]
		}
	}

	loop(0)
	return results
}

File: java-problems/src/main/java/org/whsv26/leetcode/Combinations.java
package org.whsv26.leetcode;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class Combinations {

    public static void main(String[] args) {
        var solutions = new Solution();
        var actual = solutions.combine(4, 2);
        var expected = List.of(
            List.of(1, 2),
            List.of(1, 3),
            List.of(1, 4),
            List.of(2, 3),
            List.of(2, 4),
            List.of(3, 4)
        );
        assert expected.equals(actual);
    }

    static class Solution {
        public List<List<Integer>> combine(int n, int k) {
            List<List<Integer>> res = new ArrayList<>();
            loop(new Stack<>(), 0, n, k, res);
            return res;
        }

        private void loop(Stack<Integer> stack, int from, int to, int k, List<List<Integer>> res) {
            if (stack.size() == k) {
                res.add(stack.stream().toList());
                return;
            }

            for (int i = from; i < to; i++) {
                var available = to - i;
                var requested = k - stack.size();
                if (requested <= available) {
                    stack.push(i + 1);
                    loop(stack, i + 1, to, k, res);
                    stack.pop();
                }
            }
        }
    }
}